private-templates-service/client/src/components/Designer/Designer.tsx

```diff
@@ -12,45 +14,85 @@ import * as ACDesigner from 'adaptivecards-designer';
+    sampleDataJSON: state.currentTemplate.sampleDataJSON
   };
 };
 
+const mapDispatchToProps = (dispatch: any) => {
+  return {
+    updateTemplate: (templateID: string, templateJSON: string, templateName: string, sampleDataJSON: string) => {
+      dispatch(updateTemplate(templateID, templateJSON, templateName, sampleDataJSON));
+    }
+  }
+}
+
 interface DesignerProps {
   isAuthenticated: boolean;
   user?: UserType;
+  templateID: string;
+  templateJSON: string;
+  templateName: string;
+  sampleDataJSON: string;
+  updateTemplate: (templateID: string, templateJSON: string, templateName: string, sampleDataJSON: string) => any;
 }
 
-const Designer = (props: DesignerProps) => {
+let designer: ACDesigner.CardDesigner;
+
+class Designer extends React.Component<DesignerProps> {
+
+  componentWillMount() {
+    ACDesigner.GlobalSettings.enableDataBindingSupport = true;
+    ACDesigner.GlobalSettings.showSampleDataEditorToolbox = true;
+
+    ACDesigner.CardDesigner.onProcessMarkdown = (text: string, result: { didProcess: boolean, outputHtml?: string }) => {
+      result.outputHtml = new markdownit().render(text);
+      result.didProcess = true;
+    }
+
+    designer = initDesigner();
+
+    let publishButton = new ACDesigner.ToolbarButton("publishButton", "Publish", "", (sender) => (alert("Published!")));
+    publishButton.separator = true;
+    designer.toolbar.insertElementAfter(publishButton, ACDesigner.CardDesigner.ToolbarCommands.TogglePreview);
 
-  useEffect(() => {
+    let saveButton = new ACDesigner.ToolbarButton("saveButton", "Save", "", (sender) => (onSave(designer, this.props)));
+    saveButton.separator = true;
+    designer.toolbar.insertElementAfter(saveButton, ACDesigner.CardDesigner.ToolbarCommands.TogglePreview);
+
+    designer.sampleData = "";
+  }
+
+  componentDidMount() {
     const element = document.getElementById("designer-container");
     if (element) {
       designer.attachTo(element);
     }
 
     designer.monacoModuleLoaded(monaco);
 
-    getCardData(designer);
-  }, [])
-
-  let history = useHistory();
+    console.log("id in designer: " + this.props.templateID);
+  }
 
-  ACDesigner.GlobalSettings.enableDataBindingSupport = true;
-  ACDesigner.GlobalSettings.showSampleDataEditorToolbox = true;
+  render() {
 
-  ACDesigner.CardDesigner.onProcessMarkdown = (text: string, result: { didProcess: boolean, outputHtml?: string }) => {
-    result.outputHtml = new markdownit().render(text);
-    result.didProcess = true;
+    return (
+      <div id="designer-container" dangerouslySetInnerHTML={{ __html: "dangerouslySetACDesigner" }}></div>
+    )
   }
 
-  let designer = initDesigner();
-
-  let closeButton = new ACDesigner.ToolbarButton("closeButton", "Close", "", (sender) => (history.goBack()));
-  closeButton.separator = true;
-  designer.toolbar.insertElementAfter(closeButton, ACDesigner.CardDesigner.ToolbarCommands.TogglePreview);
+  shouldComponentUpdate() {
```

Comment:
Is this needed? (I looked it up quickly, and render won't be called if it returns false...)

Reply:
Yeah it is needed. Basically what was happening is that every time a prop of this Designer was updated, the entire designer would render all over again. So whenever I hit "Save" on a template it would update the templateJSON prop of Designer, triggering a re-render of the component. But when the Designer component re-rendered, the template in the designer would be changed to the default template and not the template I was working on. By having this method here Designer will never need to re-render after it is first rendered, which helps with performance and ensuring "Save" works correctly.

Reply:
Why was the save button causing it to render again?

Reply:
Updating template in redux -> updating props -> re render?

Reply:
Yes exactly that. Save caused the templateJSON field in redux to update, which causes props of Designer to update, which causes a re-render